### PR TITLE
Add syncfree Adam and AdamW optimizers

### PR DESCRIPTION
**Pull request by ymwangg**
This PR adds syncfree Adam/AdamW optimizers and does some minor refactoring on SGD optimizer. We saw ~26% speedup using syncfree Adam/AdamW in AMP mode using the `test_train_mp_imagenet_amp.py` script.

The following is the throughput comparison chart collected on Nvidia Tesla V100 GPU.

| Optimizer | original | syncfree | speedup |
| --------- | ------- | -------- | -------- |
| SGD         | 965.41 | 1152.50 | 1.19 |
| Adam       | 869.79 | 1093.69 | 1.26 |
| AdamW    | 873.55 | 1100.79 | 1.26 |

Co-authored by `Ricky Das <ritwikdas54@gmail.com>` @codeislife99

### CODE REVIEW

**Comment by JackCaoG** [comment]
Thanks @ymwangg I will review it soon. Any chance you can share more detial data? Like how manny gpu you guys tested and what's the throughput?

**Comment by ymwangg** [comment]
@JackCaoG I added some throughput data. Let me know if you have any questions.

**Comment by JackCaoG** [comment]
Sorry for the delay, will try to take a look this week.

File: torch_xla/csrc/xla_lower_util.cpp [hunk 1]
```diff
@@ -866,6 +866,72 @@ std::vector<xla::XlaOp> BuildSgdOptimizerStep(
...
+  xla::PrimitiveType type = XlaHelpers::ShapeOfXlaOp(param).element_type();
+  xla::XlaOp one = xla::One(param.builder(), type);
+  xla::XlaOp zero = xla::Zero(param.builder(), type);
+
+  xla::XlaOp found_inf_cond = xla::Ne(found_inf, zero);
+
+  // update step counter first
+  xla::XlaOp not_found_inf =
+      xla::ConvertElementType(xla::Not(found_inf_cond), type);
+  xla::XlaOp new_step = step + not_found_inf;
```
Comment (JackCaoG) [inline torch_xla/csrc/xla_lower_util.cpp:889]: Could you add a comment here to explain this `new_step` computation?

**Review by JackCaoG — COMMENTED** [review]
Most lgtm. One minor question

File: torch_xla/csrc/xla_lower_util.cpp (hunk 1 above)
Comment (ymwangg) [reply torch_xla/csrc/xla_lower_util.cpp:893]: Done.

**Comment by ymwangg** [comment]
I noticed the CI script has some issues and the optimizer related tests were not run. I'll try to fix it in this PR. Not sure why it did not exit with an error.
```sh
++ command -v nvidia-smi
+ '[' -x /usr/bin/nvidia-smi ']'
+ '[' -d './amp/torch_xla/amp/syncfree]'
.circleci/common.sh: line 123: [: missing `]
```

**Comment by JackCaoG** [comment]
Thanks! LGTM. @ymwangg Could you rebase this pr? I will approve once all test passed.

**Comment by ymwangg** [comment]
@JackCaoG Done. I saw OOM error in the imagenet test case so I disabled the `test_train_mp_imagenet_amp.py --fake_data --num_epochs=1` test. I'm not sure why it's included in CI but it will take more than 1 hour to finish.

**Review by JackCaoG — APPROVED** [review]
Thanks @ymwangg !

**Pull request merged**